docs/quickstart.rst

```diff
@@ -8,51 +8,156 @@ up pg_auto_failover to replicate data between them. We’ll simulate failure in
+on-premise network. We'll use four virtual machines: a primary database, a
+secondary database, a monitor, and an "application." The monitor watches the
+other nodes’ health, manages global state, and assigns nodes their roles.
+
+.. _quickstart_network:
+
+Create virtual network
+----------------------
+
+Our database machines need to talk to each other and to the monitor node, so
+let's create a virtual network.
+
+.. code-block:: bash
+
+   az group create \
+       --name ha-demo \
+       --location eastus
+
+   az network vnet create \
+       --resource-group ha-demo \
+       --name ha-demo-net \
+       --address-prefix 10.0.0.0/16
+
+We need to open ports 5432 (Postgres) and 22 (SSH) between the machines, and
+also give ourselves address from our remote IP. We'll do this with a network
+security group and a subnet.
+
+.. code-block:: bash
+
+   az network nsg create \
+       --resource-group ha-demo \
+       --name ha-demo-nsg
+
+   az network nsg rule create \
+       --resource-group ha-demo \
+       --nsg-name ha-demo-nsg \
+       --name ha-demo-ssh-and-pg \
+       --access allow \
+       --protocol Tcp \
+       --direction Inbound \
+       --priority 100 \
+       --source-address-prefixes `curl ifconfig.me` 10.0.1.0/24 \
+       --source-port-range "*" \
+       --destination-address-prefix "*" \
+       --destination-port-ranges 22 5432
+
+   az network vnet subnet create \
+       --resource-group ha-demo \
+       --vnet-name ha-demo-net \
+       --name ha-demo-subnet \
+       --address-prefixes 10.0.1.0/24 \
+       --network-security-group ha-demo-nsg
+
+Finally add four virtual machines (ha-demo-a, ha-demo-b, ha-demo-monitor, and
+ha-demo-app). For speed we background the ``az vm create`` processes and run
+them in parallel:
+
+.. code-block:: bash
+
+   # allow VMs to resolve each other via internal DNS
+   az network private-dns zone create \
+       --resource-group ha-demo \
+       --name ha-demo.local
+   az network private-dns link vnet create \
+       --name ha-demo-zone-link \
+       --resource-group ha-demo \
+       --virtual-network ha-demo-net \
+       --zone-name ha-demo.local \
+       --registration-enabled true
+
+   # create VMs in parallel
+   for node in monitor a b app
+   do
+   az vm create \
+       --resource-group ha-demo \
+       --name ha-demo-${node} \
+       --vnet-name ha-demo-net \
+       --subnet ha-demo-subnet \
+       --nsg ha-demo-nsg \
+       --public-ip-address ha-demo-${node}-ip \
+       --image debian \
+       --admin-username ha-admin \
+       --generate-ssh-keys &
+   done
+   wait
+
+To make it easier to SSH into these VMs in future steps, let's make a shell
+function to retrieve their IP addresses:
+
+.. code-block:: bash
+
+  # run this in your local shell as well
+
+  vm_ip () {
+    az vm list-ip-addresses -g ha-demo -n ha-demo-$1 -o tsv \
+      --query '[] [] .virtualMachine.network.publicIpAddresses[0].ipAddress'
+  }
 
```

Comment:
I think now would be a good time to pause and show what we've done, the 4 VMs that are running in a network with known names etc.

Reply:
Maybe something like this would get the point across (what we did so far) for a quick review?

```
~ az resource list --query "[?resourceGroup=='ha-demo-dim'].{ name: name, flavor: kind, resourceType: type, region: location }" --output table
Name                             ResourceType                                           Region
-------------------------------  -----------------------------------------------------  --------
ha-demo-a                        Microsoft.Compute/virtualMachines                      eastus
ha-demo-app                      Microsoft.Compute/virtualMachines                      eastus
ha-demo-b                        Microsoft.Compute/virtualMachines                      eastus
ha-demo-monitor                  Microsoft.Compute/virtualMachines                      eastus
ha-demo-appVMNic                 Microsoft.Network/networkInterfaces                    eastus
ha-demo-aVMNic                   Microsoft.Network/networkInterfaces                    eastus
ha-demo-bVMNic                   Microsoft.Network/networkInterfaces                    eastus
ha-demo-monitorVMNic             Microsoft.Network/networkInterfaces                    eastus
ha-demo-nsg                      Microsoft.Network/networkSecurityGroups                eastus
ha-demo.local                    Microsoft.Network/privateDnsZones                      global
ha-demo.local/ha-demo-zone-link  Microsoft.Network/privateDnsZones/virtualNetworkLinks  global
ha-demo-a-ip                     Microsoft.Network/publicIPAddresses                    eastus
ha-demo-app-ip                   Microsoft.Network/publicIPAddresses                    eastus
ha-demo-b-ip                     Microsoft.Network/publicIPAddresses                    eastus
ha-demo-monitor-ip               Microsoft.Network/publicIPAddresses                    eastus
ha-demo-net                      Microsoft.Network/virtualNetworks                      eastus
```